src/import/trace-event-json.ts

```diff
@@ -0,0 +1,229 @@
+import {ProfileBuilderInfo, Sample, StackFrame, TraceEventJsonObject} from './trace-event'
+
+/**
+ * The chrome json trace event spec only specifies name and category
+ * as required stack frame properties
+ *
+ * https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.b4y98p32171
+ */
+function frameInfoForEvent({name, category}: StackFrame): FrameInfo {
+  return {
+    key: `${name}:${category}`,
+    name: name,
+  }
+}
+
+/**
+ * Initialization function to enable O(1) access to the set of active nodes in the stack by node ID.
+ */
+function getActiveNodeArrays(profile: TraceEventJsonObject): Map<number, number[]> {
+  const map: Map<number, number[]> = new Map<number, number[]>()
+
+  // Given a nodeId, `getActiveNodes` gets all the parent nodes in reversed call order
+  const getActiveNodes = (id: number): number[] => {
+    if (map.has(id)) return map.get(id) || []
+
+    const node = profile.stackFrames[id]
+    if (!node) throw new Error(`No such node ${id}`)
+    if (node.parent) {
+      const array = getActiveNodes(node.parent).concat([id])
+      map.set(id, array)
+      return array
+    } else {
+      return [id]
+    }
+  }
+
+  Object.keys(profile.stackFrames).forEach(nodeId => {
+    const id = Number(nodeId)
+    map.set(id, getActiveNodes(id))
+  })
+
+  return map
+}
+
+/**
+ * Returns an array containing the time difference in microseconds between the previous
+ * sample and the current sample
+ */
+function getTimeDeltas(contents: TraceEventJsonObject) {
+  const timeDeltas: number[] = []
+  let lastTimeStamp = Number(contents.samples[0].ts)
+
+  contents.samples.forEach((sample: Sample, idx: number) => {
+    if (idx === 0) {
+      timeDeltas.push(0)
+    } else {
+      const timeDiff = Number(sample.ts) - lastTimeStamp
+      lastTimeStamp = Number(sample.ts)
+      timeDeltas.push(timeDiff)
+    }
+  })
+
+  return timeDeltas
+}
+
+export function constructProfileFromJsonObject(
+  contents: TraceEventJsonObject,
+  samplesForPidTid: Sample[],
+  {profileBuilder}: ProfileBuilderInfo,
+) {
+  const activeNodeArraysById = getActiveNodeArrays(contents)
+
+  /**
+   * The json object format maintains an object of stack frames where the
+   * key is the frame id and the value is the stack frame object.
+   */
+  function getFrameById(frameId: string | number): StackFrame {
+    return contents.stackFrames[String(frameId)]
+  }
+
+  /**
+   * Wrapper function to get the active nodes for a given node id. We should
+   * always have active nodes for any given node.
+   */
+  function getActiveNodeIds(nodeId: number): number[] {
+    const activeNodeIds = activeNodeArraysById.get(nodeId)
+    if (!activeNodeIds) throw new Error(`No such node ID ${nodeId}`)
+    return activeNodeIds
+  }
+
+  // We need to leave frames in the same order that we start them, so we keep a stack
+  // of frames that are currently open
+  const frameStack: StackFrame[] = []
```

Comment:
There is an opportunity to de-duplicate some of this logic i can look into in a bit